### ISSUE TITLE
Test case run() methods to use Deferreds internally

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10918]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1341|
|Type|defect|
|Created|2005-11-28 05:35:42Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1341 1341
type__defect defect
reporter__jml jml
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1133156142000000 1133156142000000
changetime__1135087599000000 1135087599000000
version__ 
owner__ 
cc__exarkun cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
In trunk TestCase.run() looks a little like:
  wait(setUp)
  wait(testMethod)
  wait(tearDown)

This is pretty lame.  Ideally, Trial should behave like a Twisted application
and asynchronous methods should just return Deferreds, rather than using
blocking necromancy.

Th branch /jml/trial-deferred-logic is the first step.  It changes the
implementation of the run() method to look something like:
  d = setUp()
  d.addCallback(testMethod)
  d.addCallback(tearDown)
  wait(d)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
In twisted/web/test/test_webclient.py, calls to reactor.iterate() after
port.stopListening() are removed.  I think the method should also be returning
the result of stopListening() now (also I don't see the point of deleting
self.port, that can probably be deleted without causing any harm).

In twisted/mail/test/test_mail.py, the tearDown method which called
reactor.iterate() probably served to reduce the likelyhood of a race going the
wrong way.  This test should really have a Deferred it is waiting on, but the
current API does not allow for this.  Removing the tearDown method may increase
the likelihood of spurious failures being reported from this test.

In twisted/trial/runner.py, ClassSuite.__init__ calls pyunit.TestSuite.__init__,
but ClassSuite subclasses TestSuite.  This didn't change in the branch, but it
should be fixed at some point anyway.

In ClassSuite._run, the test is skipped if "testCase" has no attribute
methodName.  Under what circumstances can this occur?

ClassSuite.run adds an errback to the run with a comment "swallow setUpClass
failure", but that errback will swallow errors from any point in the entire run.
 If this kind of action at a distance is desired, setUpClass should wrap
whatever error it handles and then propagates with a well-known exception class
(one not used elsewhere) and the errback should trap that particular type before
deciding it does not need to propagate.  Alternatively, the callback/errback
chain should be set up to avoid this possible problem. 
twisted/trial/unittest.py's TestCase.run() seems to have a similar problem.

SignalStateManager.save()/restore() should probably be a try/finally pair.

Also, SignalStateManager has been lifted off ClassSuite and made into a local in
 the run() method here, but in twisted/trial/unittest.py's TestCase's run()
method, it has been lifted out of locals and made into a class attribute. 
What's the goal of these contrary changes?

Aside from these concerns, the changes look like a good step in the right direction.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
> In ClassSuite._run, the test is skipped if "testCase" has no attribute 
> methodName.  Under what circumstances can this occur?

When testCase has no setUpClass or tearDownClass.  Generally occurs when tests
inherit from pyunit.TestCase.  I guess this needs to be made more explicit.

> In twisted/mail/test/test_mail.py, the tearDown method which called
> reactor.iterate() probably served to reduce the likelyhood of a race going the
> wrong way.  This test should really have a Deferred it is waiting on, but the
> current API does not allow for this.  Removing the tearDown method may 
> increase the likelihood of spurious failures being reported from this test.

Unfortunately, leaving the reactor.iterate() calls in there make it certain that
the test will fail.  Suggestions?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
From an email to TPML by exarkun:

>@@ -907,17 +900,22 @@
>         self.client.transport.loseConnection()
>         return self.p.stopListening()
>
>+    def _delayDeferred(self, time, arg=None):
>+        from twisted.internet import reactor
>+        d = defer.Deferred()
>+        reactor.callLater(time, d.callback, arg)
>+        return d
>+
>     def testNoNotification(self):
>         client = self.client
>         f = self.f
>         client.transport.write("hello")
>         w = client.transport.write
>         client.transport.loseWriteConnection()
>-        reactor.iterate()
>-        reactor.iterate()
>-        reactor.iterate()
>-        self.assertEquals(f.protocol.data, "hello")
>-        self.assertEquals(f.protocol.closed, True)
>+        d = self._delayDeferred(0.2, f.protocol)
>+        d.addCallback(lambda x : self.assertEqual(f.protocol.data, 'hello'))
>+        d.addCallback(lambda x : self.assertEqual(f.protocol.closed, True))
>+        return d
>
>     def testShutdownException(self):
>         client = self.client

The above hunk [from test_tcp -- jml] is pretty sketchy.  The real solution is
to give the protocol a Deferred which it fires at whatever point it would have
set its closed attribute to True.  As it stands, the test has a race and I can
guarantee it will be going the wrong way quite frequently.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
Moved to different issue.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
